kong/clustering/control_plane.lua

```diff
@@ -0,0 +1,534 @@
+
+          else
+            ngx_log(ngx_DEBUG, _log_prefix, "sent pong frame to data plane", log_suffix)
+          end
+
+        else -- is reconfigure
+          local previous_sync_status = sync_status
+          ok, err, sync_status = self:check_configuration_compatibility(
+                                    { dp_plugins_map = dp_plugins_map, })
+          if ok then
+            local has_update, deflated_payload, err = update_compatible_payload(self.reconfigure_payload, dp_version, log_suffix)
+            if not has_update then -- no modification, use the cached payload
+              deflated_payload = self.deflated_reconfigure_payload
+            elseif err then
+              ngx_log(ngx_WARN, "unable to update compatible payload: ", err, ", the unmodified config ",
+                                "is returned", log_suffix)
+              deflated_payload = self.deflated_reconfigure_payload
+            end
+
+            -- config update
+            local _, err = wb:send_binary(deflated_payload)
+            if err then
+              if not is_timeout(err) then
+                return nil, "unable to send updated configuration to data plane: " .. err
+              end
+
+              ngx_log(ngx_NOTICE, _log_prefix, "unable to send updated configuration to data plane: ", err, log_suffix)
+
+            else
+              ngx_log(ngx_DEBUG, _log_prefix, "sent config update to data plane", log_suffix)
+            end
+
+          else
+            ngx_log(ngx_WARN, _log_prefix, "unable to send updated configuration to data plane: ", err, log_suffix)
+            if sync_status ~= previous_sync_status then
+              update_sync_status()
+            end
+          end
+        end
+
+      elseif err ~= "timeout" then
+        return nil, "semaphore wait error: " .. err
+      end
+    end
+  end)
+
+  local ok, err, perr = ngx.thread.wait(write_thread, read_thread)
+
+  self.clients[wb] = nil
+
+  ngx.thread.kill(write_thread)
+  ngx.thread.kill(read_thread)
+
+  wb:send_close()
+
+  --TODO: should we update disconnect data plane status?
+  --sync_status = CLUSTERING_SYNC_STATUS.UNKNOWN
+  --update_sync_status()
+
+  if not ok then
+    ngx_log(ngx_ERR, _log_prefix, err, log_suffix)
+    return ngx_exit(ngx_ERROR)
+  end
+
+  if perr then
+    ngx_log(ngx_ERR, _log_prefix, perr, log_suffix)
+    return ngx_exit(ngx_ERROR)
+  end
+
+  return ngx_exit(ngx_OK)
+end
+
+
+local function push_config_loop(premature, self, push_config_semaphore, delay)
+  if premature then
+    return
+  end
+
+  local ok, err = handle_export_deflated_reconfigure_payload(self)
+  if not ok then
+    ngx_log(ngx_ERR, _log_prefix, "unable to export initial config from database: ", err)
+  end
+
+  while not exiting() do
+    local ok, err = push_config_semaphore:wait(1)
+    if exiting() then
+      return
+    end
+
+    if ok then
+      if isempty(self.clients) then
```

Comment:
It looks like we do not need a queue here. Just a counter is enough. Ref: https://github.com/Kong/kong/blob/b42d53bcc694630ddb4bdafda2c645ceab7e5823/kong/clustering/control_plane.lua#L154-L155